paddle/fluid/operators/conv_transpose_op.cc

```diff
@@ -1,449 +0,0 @@
-limitations under the License. */
-
-#include "paddle/fluid/operators/conv_transpose_op.h"
-
-#include <string>
-#include <vector>
-
-#include "paddle/fluid/framework/data_layout.h"
-#include "paddle/fluid/framework/infershape_utils.h"
-#include "paddle/fluid/framework/op_registry.h"
-#include "paddle/fluid/framework/op_version_registry.h"
-#include "paddle/phi/core/infermeta_utils.h"
-#include "paddle/phi/infermeta/backward.h"
-#include "paddle/phi/infermeta/binary.h"
-#ifdef PADDLE_WITH_MKLDNN
-#include "paddle/fluid/platform/mkldnn_helper.h"
-#endif
-
-namespace paddle {
-namespace operators {
-
-using DataLayout = phi::DataLayout;
-
-phi::KernelKey ConvTransposeOp::GetExpectedKernelType(
-    const framework::ExecutionContext& ctx) const {
-  auto data_type = OperatorWithKernel::IndicateVarDataType(ctx, "Input");
-  return phi::KernelKey(data_type, ctx.GetPlace());
-}
-
-phi::KernelKey ConvTransposeOp::GetKernelTypeForVar(
-    const std::string& var_name,
-    const phi::DenseTensor& tensor,
-    const phi::KernelKey& expected_kernel_type) const {
-#ifdef PADDLE_WITH_MKLDNN
-  // Only input require reshaping, weights and
-  // bias are having shape in NCHW order
-  if ((var_name == "Input") &&
-      (expected_kernel_type.layout() == phi::DataLayout::ONEDNN) &&
-      (tensor.layout() != phi::DataLayout::ONEDNN)) {
-    auto attrs = Attrs();
-    auto ar = paddle::framework::AttrReader(attrs);
-    const std::string data_format = ar.Get<std::string>("data_format");
-    auto dl = phi::StringToDataLayout(data_format);
-    // Some models may have intentionally set "AnyLayout" for pool
-    // op. Treat this as NCHW (default data_format value)
-    if (dl != phi::DataLayout::kAnyLayout) {
-      return phi::KernelKey(tensor.place(), dl, expected_kernel_type.dtype());
-    }
-  }
-#endif
-  return phi::KernelKey(
-      tensor.place(), tensor.layout(), expected_kernel_type.dtype());
-}
-
-void Conv2DTransposeOpMaker::Make() {
-  AddInput("Input",
-           "(Tensor) The input tensor of convolution transpose operator. "
-           "The format of input tensor is NCHW or NHWC. Where N is batch size, "
-           "C is the number of input channels, H is the height of the feature, "
-           "and W is the width of the feature.");
-  AddInput(
-      "Filter",
-      "(Tensor) The filter tensor of convolution transpose operator. "
-      "The format of the filter tensor is MCHW, where M is the number of "
-      "input feature channels, C is the number of "
-      "output feature channels,"
-      "H is the height of the filter, and W is the width of the filter. "
-      "We enforce groups number == 1 in the convolution transpose scenario.");
-  AddInput("Bias",
-           "(Tensor) Bias to be added to each output of filter application."
-           "The format of output tensor is X (one-dimensional) of size equal"
-           "to the number of output channels. Only used with MKL-DNN.")
-      .AsDispensable()
-      .AsExtra();
```

Comment:
这个 `Bias` 输入在 yaml 里没有定义：
- 我第一个 `update` commit 没有在 yaml 里没有添加这个 bias，编译能过，但是有 2 个单测失败，报错显示找不到 bias
- 然后我提交了一个 `test` commit 添加 `bias` 可选输入，但是试了一下没办法生成 `.AsExtra()`，编译能过，但是有几个单测（和第一个 commit 报错的单测不一样）失败，报错显示 bias 输入的不是 Tensor 是 List，感觉可能是不用 bias 输入的情况下也给 bias 传了值

所以这是动静不一致吗，需要和之前一样保留 `lagecy_xxx.yaml` 然后移动到 `static_xx.yaml` 吗？

Reply:
> 这个 `Bias` 输入在 yaml 里没有定义：
> 
> * 我第一个 `update` commit 没有在 yaml 里没有添加这个 bias，编译能过，但是有 2 个单测失败，报错显示找不到 bias
> * 然后我提交了一个 `test` commit 添加 `bias` 可选输入，但是试了一下没办法生成 `.AsExtra()`，编译能过，但是有几个单测（和第一个 commit 报错的单测不一样）失败，报错显示 bias 输入的不是 Tensor 是 List，感觉可能是不用 bias 输入的情况下也给 bias 传了值
> 
> 所以这是动静不一致吗，需要和之前一样保留 `lagecy_xxx.yaml` 然后移动到 `static_xx.yaml` 吗？

可以尝试将bias加入到op_compat的extra中，如果还是有问题，可能需要考虑动静不一致。

Reply:
``` yaml
- op : conv2d_transpose
  ...
  extra :
    inputs : [bias]
```
我在 `op_compat.yaml` 里试着加了，但是没办法生成 `.AsExtra()`

Reply:
> ```yaml
> - op : conv2d_transpose
>   ...
>   extra :
>     inputs : [bias]
> ```
> 
> 我在 `op_compat.yaml` 里试着加了，但是没办法生成 `.AsExtra()`

1. 可以修改一下`operator_utils.c.j2`中的`{% macro add_input(i, input, op_name) %}`：
```
{# add input, it could be duplicable or dispensable #}
{% macro add_input(i, input, op_name) %}{# inline #}
  {% set name = input["fluid_name"] %}
  {% set typename = input["typename"] %}
AddInput({{name| to_opmaker_name}}, "({{typename}}), input {{i}} of {{op_name}} op.")
  {%- if typename is vec %}

    .AsDuplicable()
  {%- endif %}
  {%- if input["optional"] %}

    .AsDispensable()
  {%- if "is_extra" in input and input["is_extra"] %}

    .AsExtra()
  {%- endif %}
{%- endmacro %}
```
如果发现功能不支持，可以尝试修改这个文件或者其他脚本
2. 修改paddle/fluid/operators/generator/generate_op.py

约320行增加代码：
```
if 'extra' in op_args and 'inputs' in op_args['extra']:
            for input_item in forward_op_item['inputs']:
                if input_item['name'] in op_args['extra']['outputs']:
                    input_item['is_extra'] = True
```

Reply:
实现了支持 extra input 的自动生成后，测试依旧报错，于是我针对 `conv2d_transpose` 和 `depthwise_conv2d_transpsoe` 这两个 op，保留了 `legacy_xxx.yaml` (无 bias 参数)，并且添加了 `static_xxx.yaml` (有 bias 参数)，本地测试能够通过。